test/TestInternalGrains/StreamingImmutabilityTestGrain.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Threading.Tasks;
+using Orleans;
+using Orleans.Streams;
+using UnitTests.GrainInterfaces;
+
+namespace UnitTests.Grains
+{
+    public class StreamingImmutabilityTestGrain : Grain, IStreamingImmutabilityTestGrain
+    {
+        private StreamImmutabilityTestObject _myObject;
+        private StreamSubscriptionHandle<StreamImmutabilityTestObject> _streamSubscriptionHandle;
+
+        public async Task SubscribeToStream(Guid guid, string providerName)
+        {
+            var stream = GetStreamProvider(providerName).GetStream<StreamImmutabilityTestObject>(guid, "Namespace");
+            _streamSubscriptionHandle = await stream.SubscribeAsync(OnNextAsync);
+        }
+
+        public async Task UnsubscribeFromStream()
+        {
+            if (_streamSubscriptionHandle != null)
+                await _streamSubscriptionHandle.UnsubscribeAsync();
+        }
+
+        public async Task ProduceTestObject(string providerName)
+        {
+            _myObject = new StreamImmutabilityTestObject() {MyString = this.GetPrimaryKey().ToString()};
+            var stream = GetStreamProvider(providerName).GetStream<StreamImmutabilityTestObject>(this.GetPrimaryKey(), "Namespace");
+            await stream.OnNextAsync(_myObject);
```

Comment:
> "two aspects" 

Yes and no.  The immutability alters exactly one behavior, weather a deep copy of an object is made at call time.  All other 'aspects' are side effects of our RPC pattern.

This means what you're testing is side effects, not the actual behavior.  As these side effects depend on a great many other factors (like grain locality) these tests could fail (in the future) due to perfectly reasonable changes to other parts of the system.

I think I now understand how you are approaching this, but I don't think testing undocumented and unsupported side effects rather than the specifics of the feature is a good approach.

This is a good improvement, so I feel bad about being such a stickler on the testing, but these test harnesses are public and are often examined by those trying to understand the system.  This test not only test side effects, but would also give readers a false impression of what immutability means.